File: ios/AirshipFrameworkProxy/LaunchDeepLinkTracker.swift
/* Copyright Airship and Contributors */

import Foundation

#if canImport(AirshipKit)
import AirshipKit
#elseif canImport(AirshipCore)
import AirshipCore
#endif

/// Tracks the deep link that launched the app from a notification tap.
///
/// The stash is consume-once and expires after `maxStashAge` so a JS reload
/// or a stale foreground tap can't replay an old launch link.
@MainActor
final class LaunchDeepLinkTracker {

    static let shared = LaunchDeepLinkTracker()

    private static let deepLinkActionKeys = ["^d", "deep_link_action"]
    private static let maxStashAge: TimeInterval = 10.0

    private var stash: (deepLink: String, date: Date)?
    private var launchResolved = false
    private var waiters: [CheckedContinuation<String?, Never>] = []

    private let dateProvider: () -> Date

    init(dateProvider: @escaping () -> Date = { Date() }) {
        self.dateProvider = dateProvider
    }

    /// Called from the notification response handler with the tapped
    /// notification's payload. A default-action tap resolves the launch,
    /// stashing the payload's deep link if present.
    func onNotificationResponse(
        userInfo: [AnyHashable: Any],
        isDefaultAction: Bool
    ) {
        guard isDefaultAction else { return }
        let deepLink = Self.deepLinkActionKeys
            .compactMap { userInfo[$0] as? String }
            .first
        if let deepLink {
            stash = (deepLink, dateProvider())
        }
        launchResolved = true
        resolveWaiters()
    }

    /// Resolves the launch without a deep link. Called once the app becomes
    /// active so normal launches resolve nil without waiting.
    func onLaunchResolved() {
        launchResolved = true
        resolveWaiters()
    }

    /// Returns the deep link that launched the app, or nil. Consumes the value.
    func takeLaunchDeepLink() async -> String? {
        if let deepLink = takeFreshStash() {
            return deepLink
        }
        if launchResolved {
            return nil
        }
        return await withCheckedContinuation { waiters.append($0) }
    }

    private func takeFreshStash() -> String? {
        guard let stash else { return nil }
        self.stash = nil
        guard dateProvider().timeIntervalSince(stash.date) <= Self.maxStashAge else {
            return nil
        }
        return stash.deepLink
    }

    private func resolveWaiters() {
        guard !waiters.isEmpty else { return }
        let result = takeFreshStash()
        let pending = waiters
        waiters = []
        pending.forEach { $0.resume(returning: result) }
    }
}

File: android/airship-framework-proxy/src/main/java/com/urbanairship/android/framework/proxy/LaunchDeepLinkTracker.kt
/* Copyright Airship and Contributors */

package com.urbanairship.android.framework.proxy

import com.urbanairship.push.PushMessage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.getAndUpdate

/**
 * Tracks the deep link that launched the app from a notification tap.
 *
 * The stash is consume-once and expires after a short staleness window so a
 * JS reload can't replay an old launch link.
 */
public class LaunchDeepLinkTracker internal constructor(
    private val clock: () -> Long = { System.currentTimeMillis() }
) {

    private data class Stash(val deepLink: String, val timeMs: Long)

    private val stash = MutableStateFlow<Stash?>(null)
    private val launchResolved = MutableStateFlow(false)

    /**
     * Called when a notification is opened, before the SDK runs the
     * notification's actions. Resolves the launch, stashing the payload's
     * deep link when the app was not already foregrounded.
     */
    internal fun onNotificationOpened(message: PushMessage, isAppForegrounded: Boolean) {
        if (!isAppForegrounded) {
            DEEP_LINK_ACTION_KEYS.firstNotNullOfOrNull { message.actions[it]?.string }?.let {
                stash.value = Stash(it, clock())
            }
        }
        launchResolved.value = true
    }

    /**
     * Resolves the launch without a deep link. Called once the app is
     * foregrounded so normal launches resolve null without waiting.
     */
    internal fun markLaunchResolved() {
        launchResolved.value = true
    }

    /**
     * Returns the deep link that launched the app, or null. Consumes the value.
     */
    public suspend fun takeLaunchDeepLink(): String? {
        take()?.let { return it }
        launchResolved.first { it }
        return take()
    }

    private fun take(): String? {
        val current = stash.getAndUpdate { null } ?: return null
        return if (clock() - current.timeMs <= MAX_STASH_AGE_MS) current.deepLink else null
    }

    public companion object {
        private val DEEP_LINK_ACTION_KEYS = listOf("^d", "deep_link_action")
        private const val MAX_STASH_AGE_MS = 10_000L

        private val sharedInstance = LaunchDeepLinkTracker()

        /**
         * Shared tracker instance.
         */
        @JvmStatic
        public fun shared(): LaunchDeepLinkTracker = sharedInstance
    }
}
